lumibot/backtesting/polygon_backtesting.py

```diff
@@ -177,15 +174,20 @@ def update_pandas_data(self, asset, quote, length, timestep, start_dt=None):
             logging.error(traceback.format_exc())
             raise Exception("Error getting data from Polygon") from e
 
-        if df is None:
-            return None
+        if (df is None) or df.empty:
+            return
 
-        pandas_data = []
         data = Data(asset_separated, df, timestep=ts_unit, quote=quote_asset)
-        pandas_data.append(data)
-        pandas_data_updated = self._set_pandas_data_keys(pandas_data)
+        pandas_data_update = self._set_pandas_data_keys([data])
 
-        return pandas_data_updated
```

Comment:
why is pandas_data_updated no longer being returned? are you sure this won't break anything?

Reply:
The only references to update_pandas_data in the library are in this file and it looked to be an internal method to me.  The purpose of the method seems to be about the side effect of updating PandasData.pandas_data and (sometimes) PandasData._data_store.  If update_pandas_data is part of the API I can certainly put the return values back in.

My question is why do we have both _data_store and pandas_data in the super class PandasData?  They hold nearly the same data but there is some kind of staging effect (i.e. out of sync behavior) whose top-level is PandasData.load_data.  For big backtests that doubling of RAM usage is a significant cost and I don't know what it is for (although as I mentioned elsewhere I haven't read the docs...).  If the two copies are needed and the selective updating has a purpose then I'll include that when adding a docstring to update_data_store.

Reply:
Ok that makes sense, in that case I think you're right to re-arrange it like that.

As for your question about _data_store vs pandas_data: I believe one of them is used for inputting data at the very beginning, while the other is used at the actual data store. I think you're right that we do not need both and would probably be better off removing one of them.

Reply:
Hmmm, yes.  After looking at the code some more, even though the name is update_pandas_data it seems the actual purpose is to update _data_store.

AFAICT pandas_data is only used by load_store.  It could be the idea is to do get_X calls in order to populate pandas_data before load_data.  But that doesn't seem to make a lot of sense because the only place that updates both is PolygonBacktesting.get_last_price.

How about I rename update_pandas_data to update_data_store and drop the option to update pandas_data.  That way if someone tries to upgrade and this would break them then they'll know.

Reply:
Oh, nevermind.  I had it backwards.  

The overrides PolygonBacktesting._pull_source_symbol_bars and PolygonBacktesting.get_historical_prices_between_dates call update_pandas_data and don't update _data_store.  So they rely on calling PandasData.load_data by _Strategy.__init__ to do some fixing up of dates or something related to options ("expiries"):

https://github.com/Lumiwealth/lumibot/blob/0599fd724405ba4ffc17b8a2d17f6912a1c51772/lumibot/strategies/_strategy.py#L215

https://github.com/Lumiwealth/lumibot/blob/0599fd724405ba4ffc17b8a2d17f6912a1c51772/lumibot/data_sources/pandas_data.py#L65

It seems as though those adjustments done by PandasData.load_data are only done when the backtest begins and won't happen for symbols that aren't preloaded into pandas_data.

I've only tested this change on the existing lumibot tests and my backtest which is just stocks and uses get_historical_prices().

Reply:
In order to do no harm I've made the default value of PolygonDataBacktesting.MAX_STORAGE_BYTES be None.  I'm concerned that I don't understand what's happening with the PolygonDataBacktesting use of pandas_data, _data_store, and load_data.  This change shouldn't have any effect (apart from the rename of update_pandas_data to _update_pandas_data which I expect isn't used by apps) unless the user enables it by assigning a suitable value to MAX_STORAGE_BYTES.